xarray/backends/api.py

```diff
@@ -15,34 +15,40 @@
 from ..core.combine import auto_combine
 from ..core.utils import close_on_error, is_remote_uri
 from ..core.pycompat import basestring, path_type
+from ..core.options import OPTIONS
 
 DATAARRAY_NAME = '__xarray_dataarray_name__'
 DATAARRAY_VARIABLE = '__xarray_dataarray_variable__'
 
 
+def _file_startswith(path, *args):
+    """Check that the file statswith with the sepcific byte sequence."""
+    with open(path, 'rb') as f:
+        for b in args:
+            sig = f.read(len(b))
+            print(sig, b)
+            return True if sig == b else False
+
+
 def _get_default_engine(path, allow_remote=False):
     if allow_remote and is_remote_uri(path):  # pragma: no cover
-        try:
-            import netCDF4
-            engine = 'netcdf4'
-        except ImportError:
-            try:
-                import pydap
-                engine = 'pydap'
-            except ImportError:
-                raise ValueError('netCDF4 or pydap is required for accessing '
-                                 'remote datasets via OPeNDAP')
+        for engine in OPTIONS['io_engines']:
+            if engine in ('netcdf4', 'pydap', 'h5netcdf'):
+                return engine
+        else:
+            raise ValueError('netCDF4 or pydap is required for accessing '
+                             'remote datasets via OPeNDAP')
     else:
-        try:
-            import netCDF4
-            engine = 'netcdf4'
-        except ImportError:  # pragma: no cover
-            try:
-                import scipy.io.netcdf
-                engine = 'scipy'
-            except ImportError:
-                raise ValueError('cannot read or write netCDF files without '
-                                 'netCDF4-python or scipy installed')
+        for engine in OPTIONS['io_engines']:
+            if engine in ('netcdf4', 'h5netcdf') \
+                    and _file_startswith(path, b'\x89HDF\r\n\x1a\n'):
```

Comment:
Rather than reopening files each time an engine is checked, can we do this once?

Maybe utility function like `_get_file_format(path)`  which returns either 'netcdf3', 'netcdf4' or 'unknown'?